[GAWB-769] Create a /status endpoint for Rawls

rtitle opened the pull request:
https://broadinstitute.atlassian.net/browse/GAWB-769
Sorry for the semi-large PR.

Features:
- `HealthMonitor` actor which periodically tests the following subsystems (period set to 1 minute), and stores the current state. Note all the status checks are asynchronous and don't depend on one another. I also defined a separate thread pool to use for status checks since it's not user facing.
  - Agora
  - Cromwell
  - DB
  - Google billing
  - Google buckets
  - Google genomics
  - Google groups
  - Google pubsub
  - LDAP
- Unauthenticated /status endpoint which asks `HealthMonitor` for the current status and returns it to the user.
- Unit tests for both of the above.

I tried to pick reasonable checks for each subsystem, but let me know if something looks off (especially for the Google ones).

Here's an example "ok" status (running locally):
```
rtitle@rawls[rt-gawb-769]$ curl "https://local.broadinstitute.org:20443/status"
{
  "ok": true,
  "systems": {
    "GooglePubSub": {
      "ok": true
    },
    "GoogleGenomics": {
      "ok": true
    },
    "LDAP": {
      "ok": true
    },
    "Database": {
      "ok": true
    },
    "Agora": {
      "ok": true
    },
    "GoogleGroups": {
      "ok": true
    },
    "GoogleBilling": {
      "ok": true
    },
    "Cromwell": {
      "ok": true
    },
    "GoogleBuckets": {
      "ok": true
    }
  }
}
```

Here's an example non-ok status (I hacked the Agora URL). Note the overall status is not-ok:
```
rtitle@rawls[rt-gawb-769]$ curl "https://local.broadinstitute.org:20443/status"
{
  "ok": false,
  "systems": {
    "GooglePubSub": {
      "ok": true
    },
    "GoogleGenomics": {
      "ok": true
    },
    "LDAP": {
      "ok": true
    },
    "Database": {
      "ok": true
    },
    "Agora": {
      "ok": false,
      "messages": ["Connection attempt to agora.fake.broadinstitute.org:443 failed"]
    },
    "GoogleGroups": {
      "ok": true
    },
    "GoogleBilling": {
      "ok": true
    },
    "Cromwell": {
      "ok": true
    },
    "GoogleBuckets": {
      "ok": true
    }
  }
}
```

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [x] **...and Orchestration's Swagger too!**
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] Tell your tech lead (TL) that the PR exists if they want to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

rtitle committed:
Added monitorable trait

rtitle committed:
More initial hacking:

rtitle committed:
Wired up all health checks. Needs cleanup and tests

rtitle committed:
Cleaned up HealthMonitorActor

rtitle committed:
Code cleanup

rtitle committed:
Progress on unit test

rtitle committed:
Got all tests passing

rtitle committed:
Wire up HealthMonitor to start on boot

rtitle committed:
Just some cleanup

rtitle committed:
Looks like a foldMap will do the trick

rtitle committed:
Finished StatusApiService

rtitle committed:
Minor clean up

rtitle committed:
Need to add status route

rtitle committed:
Fix test failure

rtitle committed:
Fix Agora and ldap tests

rtitle committed:
Update swagger and fix StatusApiService response code

rtitle committed:
Define generic rawlsEnumerationFormat to remove code duplication

rtitle committed:
Minor aesthetics

rtitle committed:
Fix unit test failure

rtitle committed:
Changed status messages to be optional

rtitle commented on core/src/main/resources/reference.conf:
```diff
@@ -18,6 +18,15 @@ submission-monitor-dispatcher {
+  type = Dispatcher
+  executor = "thread-pool-executor"
+  thread-pool-executor {
+    fixed-pool-size = 8
```
8 threads seems reasonable, as there are 8 subsystems.

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/JndiUserDirectoryDAO.scala:
```diff
@@ -48,6 +48,15 @@ class JndiUserDirectoryDAO(providerUrl: String, user: String, password: String,
 
+  override def listUsers(implicit executionContext: ExecutionContext): Future[List[RawlsUserSubjectId]] = withContext { ctx =>
+    val name = userDnFormat.replaceAll("cn=.*?,", "")
+    val filter = s"(&${userObjectClasses.map(oc => s"(objectclass=$oc)").mkString})"
```
LDAP craziness. Just looking at this myself, maybe it could use a comment.

helgridly replied:
yes please!

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala:
```diff
@@ -143,7 +144,8 @@ object Boot extends App with LazyLogging {
     val statisticsServiceConstructor: (UserInfo) => StatisticsService = StatisticsService.constructor(slickDataSource, gcsDAO, userDirDAO)
-    val methodRepoDAO = new HttpMethodRepoDAO(conf.getConfig("methodrepo").getString("server"))
+    val agoraConfig = conf.getConfig("methodrepo")
+    val methodRepoDAO = new HttpMethodRepoDAO(agoraConfig.getString("server"), agoraConfig.getString("path"))
```
I split the Agora URL config into 2 components:
```
methodrepo {
  server = "https://agora.dsde-dev.broadinstitute.org:443"
  path = "/api/v1"
}
```
This is because the Agora status endpoint just lives at /status, so I need to be able to get the base URL. I think this means an update to https://github.com/broadinstitute/firecloud-develop, I'll open a PR for that shortly.

rtitle replied:
Done https://github.com/broadinstitute/firecloud-develop/pull/456

rtitle commented:
@helgridly I think you're going to be the lucky reviewer on this. Let me know what you think.

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala:
```diff
@@ -227,6 +228,16 @@ case class SubmissionWorkflowStatusResponse(
+  ok: Boolean,
+  systems: Map[Subsystem, SubsystemStatus]
+)
+
```
Why are these in `ExecutionModel`? If you put them somewhere in the `model/` folder you can reuse them in Orchestration by republishing the rawlsModel library, too.

rtitle replied:
No particular reason. Good tip, will do.

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala:
```diff
@@ -238,21 +249,19 @@ class ExecutionJsonSupport extends JsonSupport {
-    override def read(json: JsValue): WorkflowStatus = json match {
-      case JsString(name) => WorkflowStatuses.withName(name)
-      case x => throw new DeserializationException("invalid value: " + x)
+  def rawlsEnumerationFormat[T <: RawlsEnumeration[T]](construct: String => T): RootJsonFormat[T] = {
```
cool, thanks!

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/HealthMonitor.scala:
```diff
@@ -0,0 +1,238 @@
+  // Actor API:
+
+  /** Triggers subsystem checking */
+  case object CheckAll
```
We usually make the case classes/objects we use for Actor message extend a common base class per service... but having looked into it, I have no idea why. We don't use the base class anywhere.

rtitle replied:
Yeah I don't think it's necessary for Akka but I could do that to follow the convention.

helgridly replied:
Let's be consistent for now, and decide if we want to be terse later.

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/HealthMonitor.scala:
```diff
@@ -0,0 +1,238 @@
+      (GoogleGroups, checkGoogleGroups),
+      (GooglePubSub, checkGooglePubsub),
+      (LDAP, checkLDAP)
+    ).foreach(processSubsystemResult _ tupled)
```
I'm assuming from the context that `processSubsystemResult _ tupled` unpacks the tuple and passes it to the function `processSubsystemResult`, but I'm struggling to parse it in my mind. `processSubsystemResult( _.tupled )` ?

rtitle replied:
Actually `tupled` is a method on `scala.Function2` which returns a function which takes the parameters as a tuple (instead of individual args).
See: http://www.scala-lang.org/api/2.9.2/scala/Function2.html

For example:
```
scala> def foo(a: Int, b: Int) = a + b
foo: (a: Int, b: Int)Int

scala> :t foo _
(Int, Int) => Int

scala> :t (foo _).tupled
((Int, Int)) => Int

// not relevant here, just noting :)
scala> :t (foo _).curried
Int => (Int => Int)
```

So with parens this would be written as 
```
.foreach((processSubsystemResult _).tupled)
```
Maybe that is more clear...

helgridly replied:
Isn't that more like `(processSubsystemResult.tupled) _`, then?

helgridly replied:
Oh, I'm getting confused about what _ means. `processSubsystemResult _` refers to the function itself; _ isn't a stand-in for the mapped-thing here.

rtitle replied:
Scala makes you append `_` to denote a partially applied function. So that ^ doesn't compile, but this does:
```
.foreach((processSubsystemResult _).tupled(_))
```
Note there are different meanings of `_` in that, because Scala.

I could also just scrap the whole thing and do:
```
.foreach { case (subsystem, res) => processSubsystemResult(subsystem,res) }
```

helgridly replied:
Why not just make `processSubsystemResult` take a tuple?

rtitle replied:
Yeah that sounds good :)

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/HealthMonitor.scala:
```diff
@@ -0,0 +1,238 @@
+    * Contains each subsystem status along with a timestamp of when the entry was made.
+    * Initialized with unknown status.
+    */
+  var data: Map[Subsystem, (SubsystemStatus, Long)] = {
```
gasp! a `var`!

rtitle replied:
...but it's ok because it's akka and I'm not exposing the state to any asynchronous operations (futures). I've also seen context.become used to evolve actor state which lets you use immutable data structures, but that strikes me as overly complex and not really necessary (no strong preference though).

helgridly replied:
The FP ancestors spin in their graves!

helgridly replied:
(it's fine)

geoffjentry replied:
You should make it private though even though technically no one should be able to get at it since it's all actorrefs. Over in cromwell we generally prefer FSM to context.become but it's all basically the same.

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/HealthMonitor.scala:
```diff
@@ -0,0 +1,238 @@
+
+  private def getCurrentStatus: StatusCheckResponse = {
+    val now = System.currentTimeMillis()
+    // Convert any expired statuses to unknown
```
When would this actually happen in practice? The scheduler re-checks minutely. If Rawls returns `Unknown`, is that because it's broken? :)

rtitle replied:
Hm, I was thinking about the case where a subsystem check hangs and doesn't return, but I'm enforcing a timeout on the futures so the future should fail if not complete after 1 minute. I guess this _shouldn't_ happen in practice; but it might be good as a defensive check in case our status checks stop running for any reason. (Maybe the akka scheduler dies? Not sure if that can happen.)

helgridly replied:
`Unknown` turns into an HTTP 500 anyway, so as far as the things pinging `/status` are concerned, it's still a failure. Fine by me, I guess.

helgridly commented:
A few comments to address but overall a 👍

helgridly commented:
Oh, and please use squash-and-merge when you do hit it. I think you missed it in firecloud-ui recently.

rtitle commented:
Thanks! Whoops sorry about firecloud-ui. Will address the comments.

rtitle committed:
Code review comments:
- comment LDAP stuff a little more
- move new status case classes into model/ so they can be published to Orch
- Minor cleanup of HealthMonitor

rtitle committed:
Re-add mockito dependency after rebase

rtitle commented:
Applied code review comments.

dvoet commented:
I was just passing by... the asynchronous stuff is great and all, but it seems overkill to me. Why don't we just go get status fresh every time we are asked?

rtitle commented:
A couple reasons:

1. The /status endpoint is unauthenticated - I didn't want each call to /status to trigger DB queries, Google calls, etc. It opens us up to DDoS.
2. I want the /status endpoint to be reliable. If there _is_ a problem with the DB (for example) that is causing queries to hang and time out after a minute, I don't want to tie that behavior to /status. Instead, the call to /status will return quickly and report that there is a problem with the DB (but not other subsystems).

I think a little bit of asynchronicity is a good thing in this case because it decouples the checks themselves (which might fail) with the request for status (which should be reliable). For example https://dev.twitter.com/overview/status doesn't check its subsystems in real-time; it monitors them asynchronously and reports the results when asked.

dvoet commented:
excellent, can you stick those points in a comment in the code? maybe in the scaladoc for HealthMonitor?

rtitle commented:
Will do.

rtitle committed:
Added scaladoc to HealthMonitor actor. Hope I didn't go overboard
